Add an Ethash Fake sealing option for `—dev` mode, using `—dev.ethash`

ziogaschr opened the pull request:
Implement the requested feature at #188.

ziogaschr committed:
cmd/geth, miner, params: add an Ethash Fake sealing option for `—dev` mode using `—dev.ethash`

Implement requested feature at issue #188

ziogaschr commented on cmd/utils/flags.go:
```diff
@@ -187,7 +187,11 @@ var (
+		Usage: "Block period for proof-of-authority network to use in developer mode (0 = mine only if transaction pending)",
+	}
+	DeveloperEthashFlag = cli.BoolFlag{
+		Name:  "dev.ethash",
```
@meowsbits do you think `--dev.pow` makes more sense.

Also, I was thinking if anyone might want to test a flow having the ethash and not the fake one. In this case, we can allow this cli param to pass a value like `ethash`, `fakePoW`.

meowsbits replied:
Oh yea, interesting line of thought. Following this... What happens with `--miner.threads` values? With my first implementation, this value was co-opted to configure the emission rate.

I'm remembering now that `--fakepow` is an available flag... since that _can_/_could_ be used in conjunction with `--dev.[ethash|pow]`, maybe...

`--dev.[ethash|pow]` enables `--fakepow` by default, since not requiring actual CPU dedication seems usually desireable, making `--miner.threads` configure the emission rate in seconds. This could be overridden with `--fakepow=false`, where `--miner.threads` would revert to configuring the number of dedicated CPUs. 

In general, if we can pick sane defaults that a constrained `--dev.[ethash|pow]` would configure, and then allow adjacent/existing flags to handle further, overriding configuration, then I think we can avoid a rabbit hole of a plethora of development modes (which are really just a bundle of configuration options (flag and genesis)).

ziogaschr replied:
Great ideas. So this will leave us with:

- `--dev.[ethash|pow]`
- `--fakepow`
- `--miner.threads`

Which then can be used like:

- `--dev.[ethash|pow] --miner.threads 1`, this will use ethash
- `--dev.[ethash|pow] --fakepow --miner.threads 1`, this will fakepow (we should document this nicely in the `--dev.[ethash|pow]`cli arg)

where `--miner.threads` if not set will default to:

1. follow the logic used on miner
2. set to **1** if above logic is complicated and coupled to a module

---
What do you think @meowsbits

BelfordZ replied:
Have I misunderstood, or is `--dev.[ethash|pow]` supposed to be `--dev.[ethash|poa]`?

ziogaschr replied:
By default you will use `--dev` which will enable PoA. If you prefer to have a PoW chain then you will use `--dev --dev.[ethash|pow]`.

Do you have other preferences/recommendations @BelfordZ?

meowsbits replied:
We (I?) were using `[ethash|pow]` to signal that the term was undecided (and that they're basically equivalent).  Leaning to `pow` now for Etchash vs. Ethash considerations.

ziogaschr committed:
flags: remane `isPoWNetwork` to `useEthash`

missed on last commit

ziogaschr commented on params/genesis.go:
```diff
@@ -86,14 +87,19 @@ func DefaultYoloV1GenesisBlock() *genesisT.Genesis {
-	config.Clique.Period = period
+	var config *goethereum.ChainConfig
+	if useEthash {
+		config = AllEthashProtocolChanges
```
For the `--dev` mode to work automatically from the command line, the user has to pass `--miner.threads > 0`, I was thinking that in the case of developer chain we might want to fix the auto-mining logic by doing this check:

```
if miner.threads == 0
then setThreads = 1
```

meowsbits replied:
This seems reasonable. The whole idea of the `--dev` mode is provide a practical configuration bundle.

ziogaschr replied:
Resolved by 28f9f70

meowsbits commented on consensus/ethash/sealer.go:
```diff
@@ -46,18 +47,53 @@ var (
+// Note that this is a hacky way to use ethash.threads since
+// lower values will yield faster blocks, but it saves having
+// to add or modify any more code than necessary.
+func (ethash *Ethash) makeFakeDelay() float64 {
```
@ziogaschr What do you think about breaking this into two PRs:

- an add `fakeDelay` feature which creates a fake-PoW feature that emits blocks following this distribution (no CPUs and configurable, real-world modeling rate)

- add the `dev.[ethash|pow]` flag which would activate this configuration

ziogaschr replied:
Sounds reasonable. Do we want to re-open #180 and check if it's mergable? Then change this one to only include the `--dev.[ethash.pow]` flag?

meowsbits replied:
Yea, SGTM. Though I think we'll want to modify 180 a bit to use a dedicated Ethash `Mode` instead of co-opting/re-engineering the existing `ethash.config.PowMode == ModeFake || ethash.config.PowMode == ModeFullFake`.

meowsbits commented on consensus/ethash/sealer.go:
```diff
@@ -46,18 +47,53 @@ var (
+// lower values will yield faster blocks, but it saves having
+// to add or modify any more code than necessary.
+func (ethash *Ethash) makeFakeDelay() float64 {
+	p := distuv.Poisson{
```
Also we may want to rethink the "fake delay" language. IIRC there is a "fake delayer" for testing which basically adds a `time.Sleep` for block verification (but without actually running verification). Disambiguation here may be good.

ziogaschr replied:
hmm, this simulates something similar to the difficulty. So will it make sense to name it **similar** to `makeFakeDifficultySimulation`?

meowsbits replied:
So I think we can use the word "Fake" to mean that `mix` and `nonce` don't need to be legit (are skipped during header verification). The "delay" part is the tricky part.

- https://github.com/etclabscore/core-geth/blob/master/consensus/ethash/ethash.go#L622. Here, "delay" means delay on the verification side. This is only used once, in a test, delaying for one `time.Millisecond`.
- The "delay" 180 proposes is at the _emission_ process, and is intended to model real-world emission pattern (ie Poisson).  Maybe a new `Mode` `Ethash#ModePoissonFake` is in order?

ziogaschr replied:
I am working on adding a new `Mode: Ethash#ModePoissonFake` on #180. The only question so far, is how will this mode gets enabled @meowsbits?

1. not clean, when `--dev --dev.[ethash|pow] --fakepow` instead of `Ethash#ModeFake` always use `Ethash#ModePoissonFake`. No other way to enable the Poisson mode.
2. introduce a new cli flag `--fakepoissonpow`.
3. change `--fakepow` to string arg accepting following modes [fake|poisson|full].

ziogaschr replied:
A `Mode Ethash#ModePoissonFake` has been added at #180.
The enablement of this mode has to be decided, and it will probably happen in this PR ([related comment](https://github.com/etclabscore/core-geth/pull/180#discussion_r533712441)).

meowsbits replied:
Yea. Tricky one. We're into what seems like sub-flag territory. Dev mode, with pick-your-poison consensus style; and fakepow mode, with pick-your-poisson emission style. Both have existing defaults. 

Here's what I suggest, for the sake of keeping the CLI API fairly stable, and using consistent namespacing.

- `--dev` behavior is unchanged. Uses Clique.
- `--dev.pow`. A new flag. A new mode. Incompatible with `--dev`.
- `--fakepow`. Behavior is added so that this flag actually does something when `geth` runs. (_ie_ the `SetEthash` addition). Otherwise behavior is unchanged.
- `--fakepow.poisson`. A new flag. A new fake-consensus type. Incompatible with `--fakepow`, but available and useable in the same places and ways.

This leaves the decision What ethash mode to use as a default for `--dev.pow`? I am suggesting to use not-fake PoW as the default, but, as @BelfordZ suggested, toggle the equivalent of `--mine --miner.threads=1`. The `--fakepow[.poisson]` flag(s) would be _applicable_, just not be enabled by default. 

This keeps the mode as "simple" as possible while also keeping options open for the user.

ziogaschr replied:
@meowsbits all the above have been added. The only consideration is that `--dev.pow --fakepow` doesn't sealing as I reverted your changes at #180 to the original code from go-ethereum and added your changes only for the new `ModePoissonFake`. Check [here](https://github.com/etclabscore/core-geth/blob/c3b5cb551ab95de3832a5e5e6808296458fc0fb1/consensus/ethash/sealer.go#L70-L106).

I think I misunderstood your intentions on #180. Do you want me to bring back your change so as sealing works for `--fakepow`?

meowsbits replied:
> `--dev.pow --fakepow` doesn't sealing

Does this mean that it won't automatically start mining blocks? If so, then yes, I'd suggest we change that it does start automatically mining. Or, I'm not sure what you mean by sealing doesn't work for `fakepow`...?

ziogaschr replied:
Updating here based on our discussion.

`--fakepow` is used only for the verify process and disables sealing of blocks (actually it create dump blocks), this should not be changed as it will affect the actual use case of `--fakepow`. For this reason `--dev.pow --fakepow` will not be supported as implemented https://github.com/etclabscore/core-geth/pull/243/commits/6eb9ce9e7b0ef097880deb3b65b616959a549077.

The supported modes will be:

1. `--dev.pow`: starting an Ethash Ephemeral chain
2. `--dev.pow --fakepow.poisson`: starting an Ethash Ephemeral chain, without verification of blocks and with mining delay (Poisson) based on --miner.threads

meowsbits reviewed (dismissed):
Requesting changes rel https://github.com/etclabscore/core-geth/pull/243/files#r531618009, where I think this should be 2 change sets. Open to be argued otherwise if opinions differ.

ziogaschr committed:
cmd/geth, cmd/utils/flags: rename `—dev.ethash` to `—dev.pow`

ziogaschr committed:
cmd/geth: when `—dev` and no `—miner.threads` set, default to 1

ziogaschr committed:
cmd/geth, cmd/utils: make `—dev`, `—dev.period` and `—dev.pow` exclusively selected

ziogaschr committed:
cmd/utils: allow `—fakepow` to be used in geth node main runtime

meowsbits commented on cmd/geth/main.go:
```diff
@@ -510,6 +511,9 @@ func startNode(ctx *cli.Context, stack *node.Node, backend ethapi.Backend) {
 			threads = ctx.GlobalInt(utils.LegacyMinerThreadsFlag.Name)
 			log.Warn("The flag --minerthreads is deprecated and will be removed in the future, please use --miner.threads")
 		}
+		if threads == 0 && ctx.GlobalBool(utils.DeveloperFlag.Name) {
```
Instead of checking `threads == 0`, maybe should check `!ctx.GlobalIsSet(utils.MinerThreadsFlag.Name) && !ctx.GlobalIsSet(utils.LegacyMinerThreadsFlag.Name)`. Think so?

ziogaschr replied:
If you check from line 509 till here it covers flag check, so I thought it will be cleaner this way. What do you think? At the end what we want is to deal with `threads=0`, on the other hand this will not allow user forcing `--miner.threads=0`

meowsbits replied:
I would prefer to allow users the full extent of potential control of the system, even at the cost of a little code cleanliness.  So I think we should let them be able to set `miner.threads=0`. Come to think of it, can't `miner.threads=0` be used to turn off CPU mining while still allowing GPU mining via `eth_getWork`? (See https://github.com/etclabscore/core-geth/issues/207)

meowsbits commented on cmd/utils/flags.go:
```diff
@@ -192,7 +192,11 @@ var (
+	}
+	DeveloperPoWFlag = cli.BoolFlag{
+		Name:  "dev.pow",
+		Usage: "Ephemeral proof-of-work network with a pre-funded developer account, mining enabled",
```
We need an incompatibility check (with `--dev` and `--dev.period`) for this new flag. `utils` has a helper function for that somewhere.

meowsbits replied:
`CheckExclusive`

ziogaschr replied:
I've done this already (locally) :)

meowsbits commented on params/genesis.go:
```diff
@@ -86,14 +87,19 @@ func DefaultYoloV1GenesisBlock() *genesisT.Genesis {
-	config.Clique.Period = period
+	var config *goethereum.ChainConfig
+	if useEthash {
+		config = AllEthashProtocolChanges
```
Is the `AllEthashProtocolChanges` config going to be OK to use here? There are some considerations that Clique doesn't have to make for PoW, eg. difficulty (and difficulty bomb), block rewards.

ziogaschr replied:
Good point. I haven't thought of this. I thought it will work as expected based on [initial PoW->Clique change](https://github.com/ethereum/go-ethereum/pull/15323).

Block rewards work fine. I wonder if we should care about difficulty bomb in `--dev.pow]` mode, or a good use case for it

meowsbits replied:
Dev mode shouldn't ever activate the difficulty bomb, in my opinion. And since core-geth has the ECIP available to permanently disable it, I think it should do so. We'll need to double check the `difficulty` field in the genesis block, too, to make sure it is at the `params.MinimumDifficulty` (or whatever it is (that's pseudo-code ;)). 

The thought I had in mind with the block rewards was in regards to differences available for ETH vs. ETC. For example, `AllEthashProtocolChanges` will set the reward to the ETH _Istanbul_ (permanently), while ETC might, say, set it to Era 1 of [ECIP1017](https://github.com/ethereumclassic/ECIPs/blob/master/_specs/ecip-1017.md), and would then need to establish some arbitrary value for the Era Rounds length (ie probably not 500000)... maybe `math.MaxUint64`.

ziogaschr replied:
Nice one, thanks for analysing that, it really helps me to grasp the whole system. 
I will have a look at it. It will be nice if you can give some tips on it @meowsbits , so as I learn and not spending so much time.

meowsbits replied:
Sure, check out https://github.com/etclabscore/core-geth/pull/243/commits/f4530b2c1cab406bc45f35f4429d1573490d1be5

ziogaschr replied:
Thanks, very informative. EIPs used for block rewards and difficulty bomb noted. Own you a beer

ziogaschr committed:
cmd/geth, cmd/utils, consensus/ethash, eth: add `--fakepow.poisson`

ziogaschr committed:
cmd/geth: set threads=1 on developer chain mode, only when threads not explicitly set from CLI

ziogaschr committed:
cmd/utils: disallow `—fakepow` with `—dev.pow`

meowsbits committed:
params: use custom genesis&config for dev.pow

This installs an ETC-derived genesis and config
for dev.pow mode.

Block reward disinflation is disabled, so block
rewards will always be at their initial rates.

Difficulty bomb is defused from genesis, so
no need to worry about that, ever, if CPU mining.

Date: 2020-12-04 09:04:57-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main: remove fakepow.poisson flag from 'copydb' command

Since this flag is only useful for mining (block
emission), we don't need it for copydb, which just
runs the import of chain data.

Fakepow can be used to import more quickly
by cutting corners and not doing ethash field
verification, eg if you really trust the chaindata
that you're copying.

Date: 2020-12-04 09:10:42-06:00
Signed-off-by: meows <b5c6@protonmail.com>
